### ISSUE TITLE
New Orientation Transformation

### ISSUE THREAD

**Issue by mueller-franzes** [TorchIO-project/torchio#1282]
**🚀 Feature**
A new spatial transform, potentially named `ToOrientation`, that allows reorienting images to any specified target orientation (e.g., 'LAS', 'PLS', 'RAS') defined by a 3-character axis code.

**Motivation**

TorchIO currently provides `ToCanonical`, which reorients images to the closest RAS+ (Right-Anterior-Superior) orientation. While RAS+ is a common standard, particularly in research environments processing NIfTI files, other orientations are prevalent and often desired in different contexts:

1.  **Radiological Conventions:** Images derived directly from DICOM sources or used in clinical settings frequently follow an LAS+ (Left-Anterior-Superior) convention. Directly transforming to LAS+ would be highly beneficial for users working in or interfacing with these environments.
2.  **Visualization Needs:** Depending on the anatomical structure of interest or the desired slicing plane for visualization tools, orientations like PLS+ (Posterior-Left-Superior) might offer a more intuitive view for certain tasks.
3.  **Pipeline Flexibility:** Requiring users to manually implement reorientation logic using external tools like NiBabel for orientations other than RAS+ adds complexity to preprocessing pipelines.

Providing a built-in, flexible reorientation transform enhances TorchIO's utility for a broader range of medical imaging data and workflows.

**Pitch**

Introduce a new transform class, tentatively `ToOrientation(SpatialTransform)`.

1.  **Initialization:** The class constructor should accept a `target_orientation` string parameter (e.g., `transform = ToOrientation(target_orientation='LAS')`). Input validation should ensure the string is a valid 3-character orientation code (combinations of R/L, A/P, I/S). Default to "RAS"
2.  **Functionality:** The `apply_transform` method should:
    * Determine the current orientation of each image in the subject using its affine matrix (`nib.aff2axcodes`).
    * Calculate the necessary orientation transformation to convert the image from its current orientation to the `target_orientation` using `nibabel.orientations` tools (e.g., `ornt_transform`, `apply_orientation`).
    * Apply the transformation to the image's voxel data.
    * Update the image's affine matrix to reflect the new orientation.

**Alternatives**

1.  **Manual Implementation:** Users can continue to perform custom reorientation outside of TorchIO using NiBabel directly. This is less convenient and integrated than having a dedicated TorchIO transform.
2.  **Modify `ToCanonical`:** Add an optional `target_orientation` argument to `ToCanonical`. However, this might overload the meaning of "Canonical", which specifically implies RAS+ in many contexts. A new class `ToOrientation` offers better clarity regarding its function.

**Additional context**
None

**Comment by mueller-franzes** [comment]
Possible implementation:
``` 
import nibabel as nib
import numpy as np
import torch
from torchio import Subject
from torchio import SpatialTransform


class ToOrientation(SpatialTransform):
    """Reorient the data to a specified orientation.
    
    This transform reorders the voxels and modifies the affine matrix to match
    the specified orientation code.
    
    Common orientations include:
    - "RAS": Right, Anterior, Superior (canonical, neurological convention)
    - "LAS": Left, Anterior, Superior (radiological convention)
    
    See `NiBabel docs about image orientation`_ for more information.
    
    Args:
        orientation: A three-letter orientation code (e.g., "RAS", "LPS", "LAS").
        **kwargs: See :class:`~torchio.transforms.Transform` for additional
            keyword arguments.
            
    .. _NiBabel docs about image orientation: https://nipy.org/nibabel/image_orientation.html
    """
    
    def __init__(self, orientation="RAS", **kwargs):
        super().__init__(**kwargs)
        if not isinstance(orientation, str) or len(orientation) != 3:
            raise ValueError(f"Orientation must be a 3-letter string, got {orientation}")
        
        valid_codes = set('RLAPIS')
        if not all(c in valid_codes for c in orientation):
            raise ValueError(
                f"Orientation code must be composed of characters from 'RLAPIS', got {orientation}"
            )
        
        # Check for valid axis directions
        if not (('R' in orientation or 'L' in orientation) and 
                ('A' in orientation or 'P' in orientation) and 
                ('S' in orientation or 'I' in orientation)):
            raise ValueError(
                f"Orientation code must include one character for each axis direction (RL, AP, SI), got {orientation}"
            )
            
        self.orientation = orientation
    
    def apply_transform(self, subject: Subject) -> Subject:
        for image in subject.get_images(intensity_only=False):
            affine = image.affine
            current_orientation = ''.join(nib.aff2axcodes(affine))
            
            if current_orientation == self.orientation:
                continue
                
            array = image.numpy()[np.newaxis]  # (1, C, W, H, D)
            # NIfTI images should have channels in 5th dimension
            array = array.transpose(2, 3, 4, 0, 1)  # (W, H, D, 1, C)
            
            # Create a NIfTI image
            nii = nib.Nifti1Image(array, affine)
            
            # Directly transform from current orientation to target orientation
            current_ornt = nib.orientations.io_orientation(nii.affine)
            target_ornt = nib.orientations.axcodes2ornt(tuple(self.orientation))
            transform = nib.orientations.ornt_transform(current_ornt, target_ornt)
            reoriented_array = nib.orientations.apply_orientation(nii.dataobj, transform)

            # Calculate the new affine matrix reflecting the reorientation
            reoriented_affine = nii.affine @ nib.orientations.inv_ornt_aff(transform, nii.shape)
            
            # Get the reoriented data
            array = reoriented_array.copy()
            array = array.transpose(3, 4, 0, 1, 2)  # (1, C, W, H, D)
            
            # Update the image data and affine
            image.set_data(torch.as_tensor(array[0]))
            image.affine = reoriented_affine
            
        return subject
```

**Comment by fepegar** [comment]
Thanks, @mueller-franzes. I think this is a good idea. I actually implemented something like this recently for a project. I'll compare to your implementation, open a PR, and request your review, if that's ok.

**Comment by fepegar** [comment]
Out of curiosity, when would PLS+ be useful?

**Comment by mueller-franzes** [comment]
Hi @fepegar , yes sure, thanks! 
Using `PLS+` is more of a workaround when displaying axial slices with `matplotlib.imshow`. By default, the y-axis in imshow starts at 0 at the top and increases downward.  To avoid the image being displayed "upside down", the image should be aligned anterior-to-posterior (P) along the y-axis. The x-axis should be oriented from the patient's perspective, reflecting the standard radiological view, so right to left (L). The third axis is not relevant in this context, so it remains set as inferior-to-superior (S).
I hope this makes sense.

**Comment by romainVala** [comment]
Hi
I do not  understand the motivation here.

For me there is a 2 choices:
1) one wish to be robust to orientation and one use `RandomFlip` transform ( I experiment harder training)
2) one wish to have a consistent orientation 
 
but for 2) I do not understand why choosing one orientation over the other. it looks like an arbitrary choice to me, so I am good with only one (ToCanonical)

but may be your need is only for preprocessing ? (ie : When you want to write output nifti files, you want a specific orientation )

But even though I am still curious about  your initial motivation :

1) If you want to stay with the initial orientation of your nifti file (which may come from dicom convertsion) then just do not use ToCanonical, and the orientation is not changed ..

2) Viewing functionality should be independent from the nifti file orientation (that is the case for most 3D viewer, ) ... but while writing this you just reply, so I see you want a 2D slice viewer in python. (a good viewer should handel the different orientation and then let you choose an axial or sagital view ...  ok this is more work but I think it worth it

3) ok

**Comment by fepegar** [comment]
Thanks for chipping in, @romainVala. I agree with you. In my case, this is the key:

> may be your need is only for preprocessing ?

I was trying to use [CT-CLIP](https://github.com/ibrahimethemhamamci/CT-CLIP) and it turns out they trained with SLP+ images. My guess is they just didn't bother at all with orientation and assumed all their images had the same. In practice, this means we need to convert preprocess to SLP+ for that network. So [I came up with this](https://github.com/ibrahimethemhamamci/CT-CLIP/issues/41#issuecomment-2657932457):

```python
class ToSlp:
    def __call__(self, image: tio.Image) -> tio.Image:
        image = copy.deepcopy(image)

        assert image.num_channels == 1
        data = image.numpy()[0]

        current_orientation = nio.io_orientation(image.affine)
        target_orientation = nio.axcodes2ornt(("S", "L", "P"))
        transform = nio.ornt_transform(current_orientation, target_orientation)

        new_data = nio.apply_orientation(data, transform)
        new_affine = image.affine.dot(nio.inv_ornt_aff(transform, data.shape))

        image.set_data(new_data[np.newaxis].copy())
        image.affine = new_affine

        return image
```

Which can be easily made more general. You'll notice is quite similar to @mueller-franzes' code in https://github.com/TorchIO-project/torchio/issues/1282#issuecomment-2789342267. I won't have time to add this in the next few weeks. @mueller-franzes, would you like to contribute with a PR?

**Comment by mueller-franzes** [comment]
Hi,  
@romainVala thanks for the thoughtful points.
@fepegar I'd be happy to take this on and put together a PR. I’ll build on your ToSlp idea and try to make it a bit more general as discussed.

**Comment by romainVala** [comment]
I probably over-reacted a bit, because those orientation issues are annoying (always source of hidden bug).
But I agree with you it may be interesting to have a preprocessing transform to change to any given orientation
